File: .agents/skills/bio-single-cell-batch-integration/examples/harmony_integration.R
#!/usr/bin/env Rscript
# Reference: anndata 0.10+, scanpy 1.10+, scikit-learn 1.4+, scvi-tools 1.1+ | Verify API if version differs
# Batch integration with Harmony

library(Seurat)
library(harmony)
library(ggplot2)

# Example: integrate multiple 10X samples
# Assumes sample1.rds, sample2.rds, sample3.rds exist as Seurat objects

integrate_with_harmony <- function(sample_files, output_prefix = 'integrated') {
    cat('Loading samples...\n')
    samples <- lapply(sample_files, readRDS)
    names(samples) <- gsub('\\.rds$', '', basename(sample_files))

    cat('Merging samples...\n')
    merged <- merge(samples[[1]], y = samples[-1], add.cell.ids = names(samples))

    cat('Standard preprocessing...\n')
    merged <- NormalizeData(merged)
    merged <- FindVariableFeatures(merged, selection.method = 'vst', nfeatures = 2000)
    merged <- ScaleData(merged)
    merged <- RunPCA(merged, npcs = 50)

    cat('Running Harmony integration...\n')
    merged <- RunHarmony(merged, group.by.vars = 'orig.ident', dims.use = 1:30)

    cat('Post-integration analysis...\n')
    merged <- RunUMAP(merged, reduction = 'harmony', dims = 1:30)
    merged <- FindNeighbors(merged, reduction = 'harmony', dims = 1:30)
    merged <- FindClusters(merged, resolution = 0.5)

    # Visualizations
    p1 <- DimPlot(merged, reduction = 'umap', group.by = 'orig.ident') +
        ggtitle('By Sample')
    p2 <- DimPlot(merged, reduction = 'umap', label = TRUE) +
        ggtitle('By Cluster')

    pdf(paste0(output_prefix, '_umap.pdf'), width = 14, height = 6)
    print(p1 + p2)
    dev.off()

    saveRDS(merged, paste0(output_prefix, '.rds'))
    cat('Saved:', paste0(output_prefix, '.rds\n'))

    return(merged)
}

# Run if executed directly
args <- commandArgs(trailingOnly = TRUE)
if (length(args) > 0) {
    sample_files <- args
    integrate_with_harmony(sample_files)
} else {
    cat('Usage: Rscript harmony_integration.R sample1.rds sample2.rds sample3.rds\n')
}

File: .agents/skills/bio-single-cell-batch-integration/examples/harmony_integration.py
#!/usr/bin/env python3
'''Batch integration with Harmony in Scanpy'''
# Reference: anndata 0.10+, scanpy 1.10+, scikit-learn 1.4+, scvi-tools 1.1+ | Verify API if version differs

import scanpy as sc
import harmonypy as hm
import matplotlib.pyplot as plt
import sys

def integrate_with_harmony(h5ad_files, output_prefix='integrated'):
    '''Integrate multiple scRNA-seq datasets with Harmony'''
    print('Loading datasets...')
    adatas = []
    for f in h5ad_files:
        adata = sc.read_h5ad(f)
        adata.obs['batch'] = f.replace('.h5ad', '').split('/')[-1]
        adatas.append(adata)

    print('Concatenating...')
    adata = sc.concat(adatas, label='batch', keys=[a.obs['batch'].iloc[0] for a in adatas])

    print('Preprocessing...')
    sc.pp.normalize_total(adata, target_sum=1e4)
    sc.pp.log1p(adata)
    sc.pp.highly_variable_genes(adata, n_top_genes=2000, batch_key='batch')
    adata.raw = adata
    adata = adata[:, adata.var.highly_variable]
    sc.pp.scale(adata, max_value=10)
    sc.tl.pca(adata, n_comps=50)

    print('Running Harmony...')
    ho = hm.run_harmony(adata.obsm['X_pca'], adata.obs, 'batch')
    adata.obsm['X_pca_harmony'] = ho.Z_corr.T

    print('Post-integration...')
    sc.pp.neighbors(adata, use_rep='X_pca_harmony')
    sc.tl.umap(adata)
    sc.tl.leiden(adata, resolution=0.5)

    print('Plotting...')
    fig, axes = plt.subplots(1, 2, figsize=(14, 6))
    sc.pl.umap(adata, color='batch', ax=axes[0], show=False, title='By Batch')
    sc.pl.umap(adata, color='leiden', ax=axes[1], show=False, title='By Cluster')
    plt.tight_layout()
    plt.savefig(f'{output_prefix}_umap.png', dpi=150)

    adata.write(f'{output_prefix}.h5ad')
    print(f'Saved: {output_prefix}.h5ad')

    return adata

if __name__ == '__main__':
    if len(sys.argv) > 1:
        integrate_with_harmony(sys.argv[1:])
    else:
        print('Usage: python harmony_integration.py sample1.h5ad sample2.h5ad sample3.h5ad')
